tlair/src/main/scala/at/forsyte/apalache/tla/lir/TypeLevelBuilder.scala

```diff
@@ -0,0 +1,186 @@
+package at.forsyte.apalache.tla.lir
+
+import at.forsyte.apalache.tla.lir.oper.{TlaArithOper, TlaBoolOper, TlaFunOper, TlaOper}
+
+// We use shapeless for facilitating the type-level encodings of TlaExpressions
+import shapeless._
+import poly._
+
+// A (very) rough sketch at what TLA Expressions might look like when
+// we represent out typing discipline inside of Scala's type system.
+object Tla {
+  // The type of any TLA expression
+  sealed trait Exp[T] {
+    def toString: String
+  }
+
+  // Just an alias
+  type IntT = BigInt
+
+  // A value constructs a TLA expression tagged with the Scala type it carries as a payload.
+  class Value[T](value: T) extends Exp[T]
+  case class Int(value: IntT) extends Value(value)
+  case class Str(value: String) extends Value(value)
+  case class Bool(value: Boolean) extends Value(value)
+
+  // Each compound expression is reprsented by a new type
+
+  sealed trait SeqType[T]
+  case class TlaSeq[T](seq: Seq[Exp[T]]) extends Exp[SeqType[T]]
+
+  sealed trait TupleType[A, B]
+  case class Empty() extends Exp[TupleType[Unit, Unit]]
+  case class Single[A](fst: A) extends Exp[TupleType[A, Unit]]
+  case class Tuple[A, B](fst: A, snd:B) extends Exp[TupleType[A, B]]
```

Comment:
Interesting. We would have to define tuples for all sizes, similar to Scala, right?

Reply:
I think we could actually just define tuples as

```suggestion
  case class Tuple[Items <: HList](items: Items) extends Exp[TupleType[Items]]
```

That is, just as a thin wrapper around `HList`s, since those are generalized product types already :)

Reply:
This is much nicer!